src/internal.c

```diff
@@ -624,47 +629,46 @@ void CtxResourceFree(WOLFSSH_CTX* ctx)
 static void UpdateKeyID(WOLFSSH_CTX* ctx)
 {
 #ifdef WOLFSSH_CERTS
+    WOLFSSH_PVT_KEY* pvtKey;
     word32 idx;
 
-    for (idx = 0; idx < ctx->privateKeyCount &&
-                  idx < WOLFSSH_MAX_PVT_KEYS; idx++) {
-        if (ctx->cert[idx] != NULL && ctx->certSz[idx] > 0) {
-            byte keyId;
+    for (idx = 0, pvtKey = ctx->privateKey;
+            idx < ctx->privateKeyCount && idx < WOLFSSH_MAX_PVT_KEYS;
+            idx++, pvtKey++) {
+        if (pvtKey->cert != NULL && pvtKey->certSz > 0) {
             byte* der;
 
             /* matching certificate was set, convert private key id */
-            keyId = ctx->privateKeyId[idx];
-            switch (keyId) {
+            switch (pvtKey->publicKeyFmt) {
             #ifndef WOLFSSH_NO_ECDSA_SHA2_NISTP521
                 case ID_ECDSA_SHA2_NISTP521:
-                    ctx->privateKeyId[idx] = ID_X509V3_ECDSA_SHA2_NISTP521;
+                    pvtKey->publicKeyFmt = ID_X509V3_ECDSA_SHA2_NISTP521;
                     break;
             #endif
             #ifndef WOLFSSH_NO_ECDSA_SHA2_NISTP384
                 case ID_ECDSA_SHA2_NISTP384:
-                    ctx->privateKeyId[idx] = ID_X509V3_ECDSA_SHA2_NISTP384;
+                    pvtKey->publicKeyFmt = ID_X509V3_ECDSA_SHA2_NISTP384;
                     break;
             #endif
             #ifndef WOLFSSH_NO_ECDSA_SHA2_NISTP256
                 case ID_ECDSA_SHA2_NISTP256:
-                    ctx->privateKeyId[idx] = ID_X509V3_ECDSA_SHA2_NISTP256;
+                    pvtKey->publicKeyFmt = ID_X509V3_ECDSA_SHA2_NISTP256;
                     break;
             #endif
             #ifndef WOLFSSH_NO_SSH_RSA_SHA1
                 case ID_SSH_RSA:
-                    ctx->privateKeyId[idx] = ID_X509V3_SSH_RSA;
+                    pvtKey->publicKeyFmt = ID_X509V3_SSH_RSA;
                     break;
             #endif
             }
 
             /* can use the key for non X509v3 connections too */
-            der = (byte*)WMALLOC(ctx->privateKeySz[idx], ctx->heap,
-                DYNTYPE_PRIVKEY);
+            der = (byte*)WMALLOC(pvtKey->keySz, ctx->heap, DYNTYPE_PRIVKEY);
             if (der != NULL) {
                 int ret;
-                WMEMCPY(der, ctx->privateKey[idx], ctx->privateKeySz[idx]);
-                ret = SetHostPrivateKey(ctx, keyId, 1, der,
-                            ctx->privateKeySz[idx], DYNTYPE_PRIVKEY);
+                WMEMCPY(der, pvtKey->key, pvtKey->keySz);
+                ret = SetHostPrivateKey(ctx, pvtKey->publicKeyFmt, 1,
```

Comment:
This is still failing the password logon attempt to a wolfSSHd running with X509 support. The reason is that pvtKey->publicKeyFmt here has been updated to be an X509 type. This SetHostPrivateKey though should be using the initial publicKeyFmt at the start of the function which is a non X509 type. That way there is a ID_X509_* version and a ID_* version in the list.

Reply:
This isn't user authentication. This is a server authentication issue. It looks like wolfSSHd should be offering both x509v3-ecdsa-sha2-nistp256 with the certificate and ecdsa-sha2-nistp256 using the public key from the certificate.